### PR TITLE
[opt](function)Optimize the performance of the pad function under UTF-8.

### PR DESCRIPTION
**Pull request by Mryange**
## Proposed changes

1. Removed the calculation of str_index.
2. If pad is constant, calculate pad_index only once.
3. Do not insert res_chars inside the loop; instead, insert them all together after the loop completes.
```
mysql [test]>select count(lpad(Title, 100, "abc")) from hits_10m;
+--------------------------------+
| count(lpad(Title, 100, 'abc')) |
+--------------------------------+
|                       10000000 |
+--------------------------------+
1 row in set (3.97 sec)

mysql [test]>select count(lpad(Title, 100, "abc")) from hits_10m;
+--------------------------------+
| count(lpad(Title, 100, 'abc')) |
+--------------------------------+
|                       10000000 |
+--------------------------------+
1 row in set (2.87 sec)
```

### CODE REVIEW

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

**Comment by Mryange** [comment]
run buildall

File: be/src/vec/functions/function_string.h [hunk 1]
```diff
@@ -1556,86 +1556,106 @@ class FunctionStringPad : public IFunction {
...
-                        buffer.append(str_data, str_data + str_index[len]);
-                    } else {
-                        buffer.append(str_data, str_data + str_len);
-                    }
 
-                    StringOP::push_value_string(std::string_view(buffer.data(), buffer.size()), i,
-                                                res_chars, res_offsets);
+                auto [real_len, skip_chars] = simd::VStringFunctions::skip_leading_utf8(
+                        (const char*)str_data, (const char*)str_data + str_len, len);
+                if (len <= skip_chars) {
```
Comment (HappenLee) [inline be/src/vec/functions/function_string.h:1613]: seems wired. `skip_chars` seems always `<= len`

File: be/src/util/simd/vstring_function.h [hunk 2]
```diff
@@ -187,6 +187,19 @@ class VStringFunctions {
         return p;
     }
 
+    static inline std::pair<size_t, size_t> skip_leading_utf8(const char* begin, const char* end,
```
Comment (HappenLee) [inline be/src/util/simd/vstring_function.h:190]: Add comment and change the func name. we should know what the functinon do. if just one place call the func, better be a lambda

Comment (HappenLee) [inline be/src/util/simd/vstring_function.h:190]: Add comment and change the func name. we should know what the functinon do. if just one place call the func, better be a lambda

**Comment by Mryange** [comment]
run p0

File: be/src/vec/functions/function_string.h [hunk 3]
```diff
@@ -1556,86 +1556,110 @@ class FunctionStringPad : public IFunction {
...
-                        buffer.append(str_data, str_data + str_len);
-                    }
 
-                    StringOP::push_value_string(std::string_view(buffer.data(), buffer.size()), i,
-                                                res_chars, res_offsets);
+                auto [iterate_byte_len, iterate_char_len] =
+                        simd::VStringFunctions::iterate_utf8_with_limit_length(
+                                (const char*)str_data, (const char*)str_data + str_len, len);
+                // If iterate_char_len is greater than len, it indicates that the string's length exceeds len, so truncation.
+                if (iterate_char_len >= len) {
```
Comment (HappenLee) [inline be/src/vec/functions/function_string.h:1617]: not greater， only equal need to do truncate

**Comment by Mryange** [comment]
run buildall

**Review by HappenLee — APPROVED** [review]
LGTM

**Review by Gabriel39 — APPROVED** [review]

**Pull request merged**